src/hotspot/share/runtime/safepoint.cpp

```diff
@@ -603,7 +603,8 @@ void SafepointSynchronize::do_cleanup_tasks() {
   WorkerThreads* cleanup_workers = heap->safepoint_workers();
   if (cleanup_workers != nullptr) {
     // Parallel cleanup using GC provided thread pool.
-    cleanup_workers->run_task(&cleanup);
+    const uint num_workers = MIN2<uint>(SAFEPOINT_CLEANUP_NUM_TASKS, cleanup_workers->max_workers());
+    cleanup_workers->run_task(&cleanup, num_workers);
```

Comment:
If you go to the definition of `WorkerThreads::run_task(WorkerTask* task)`, then you'll notice it uses `_active_workers`. So we should be doing `MIN2(..., cleanup_workers->active_workers()` here if the current behavior is to be preserved better. I think active workers is correct in the presence of dynamic GC worker count.

Reply:
Now that I am looking at it, I wonder if we have a pre-existing performance issue here: what if active GC workers is very low for some GC-specific reason? We then run all safepoint cleanups, even for non-GC safepoints, using that small worker count? This might be the accidental reliance on the number of GC workers to match the static fraction of the number of CPUs allocated to the JVM, which is often true, but this would not hold exactly under `UseDynamicNumberOfGCThreads`.

Reply:
It is a "known" issue that if the particular task does not select its number of threads properly before executing the task (or not at all like this task) that there may be perf issues.

If this is what you want I would suggest that `active_workers()` as limit preserves the current behavior best. Otherwise the task should calculate and set its optimal number of tasks itself as it does in this change.

Not sure if setting it to `SAFEPOINT_CLEANUP_NUM_TASKS` is a good idea if some/most of this work is trivial/empty but it's probably a good first guess.

The only GC that does not seem to set safepoint_workers to something non-default seems to be Shenandoah (the stw collectors use the "main" workerthreads they also use for evacuation, so this should be maxed out mostly, and ZGC explicitly sets them to max_workers at initialization).

Reply:
@xmas92: what is the reason for this change, did you see any situation where an inappropriate amount (too few) number of threads were used? Just curious.

Reply:
BTW, as we are micro-optimizing here, then we can be quite a bit smarter about this:

```
class ParallelCleanupTask {

...
  size_t expected_workers() {
    // This worker does the light-weight tasks and guarantees progress
    size_t workers = 1;

    // If there are heavy-weight tasks pending, add more workers for them
    if (SymbolTable::needs_rehashing()) workers++;
    if (StringTable::needs_rehashing()) workers++;
    if (_do_lazy_roots)                 workers++;

    return workers;
  }
```

Reply:
@tschatzl I noticed the behaviour when I added JFR events to worker threads activity, when I was profiling generational ZGC. There we have separate WorkerThreads for each generation. And a pool a "Runtime" WorkerThreads which is only used for `CollectedHeap::safepoint_workers()`. We do not configure this and only give the VM ParallelGCThreads number of workers. 

It seems correct to respect this flag for heap inspection and heap dump, but seemed strange and unnecessary to spin up all workers when we know that there is a limit to the amount of parallel work. (Just looking at the code, the heap inspect always uses one worker, and heap dumper uses the active number of workers)

Overall my thoughts on which number to use is that:
  * **Active**: Is a bit of a strange API, because it depends on some external set state, it is the current behaviour, so maybe it should be kept and changed in a different PR.
  * **Created**: Some sort of best effort, only use as many as the system has created so far, also seems to respect `UseDynamicNumberOfGCThreads` better. 
  * **Max**: We are a task that requires X workers, we have at least X workers in `CollectedHeap::safepoint_workers()` let us use them. Sort of the most pure API, and is easier to reason about.

The idea with this PR was mostly to put an upper limit to the number of workers. So I will change to use the `active_workers`. 

Maybe `expected_workers()` is worthwhile, I have no data on this. This PR was not motivated by any performance numbers, only the missmatch between the number of workers used and the amount of work available.

Reply:
>It seems correct to respect this flag for heap inspection and heap dump, but seemed strange and unnecessary to spin up all workers when we know that there is a limit to the amount of parallel work. (Just looking at the code, the heap inspect always uses one worker, and heap dumper uses the active number of workers)

Not sure. Imho both should determine the number of threads based on some rough estimate of work too.

>[....]
>Active: Is a bit of a strange API, because it depends on some external set state, it is the current behaviour, so maybe it should be kept and changed in a different PR.

This has been long time ago added purely for convenience to not have pass it along in a sequence of parallel tasks.
I.e. in a lot of cases the "active" number is determined at start of gc (based on e.g. heap size/overall expected work), and then reduced based on the particular task like suggested in this CR.

It is true that this value should probably not be stored in `WorkerThreads`, but then the GCs would need to pass it along elsewhere.